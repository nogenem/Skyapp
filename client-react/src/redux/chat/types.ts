import { TMessageType } from '~/constants/message_types';
import { TUserStatus } from '~/constants/user_status';

interface IOtherUser {
  _id: string;
  nickname: string;
  thoughts: string;
  status: TUserStatus;
  online: boolean;
  channel_id?: string;
}

interface IOtherUsers {
  [_id: string]: IOtherUser;
}

interface IMember {
  user_id: string;
  is_adm: boolean;
  last_seen: Date;
}

interface IChannel {
  _id: string;
  name: string;
  is_group: boolean;
  members: IMember[];
  other_member_idx?: number;
  unread_msgs: number;
  lastMessage?: IMessage;
}

interface IChannels {
  [_id: string]: IChannel;
}

interface IAttachment {
  originalName: string;
  size: number;
  path: string;
  mimeType: string;
  imageDimensions?: { width: number; height: number };
}

interface IMessage {
  _id: string;
  channel_id: string;
  from_id?: string;
  body: string | IAttachment;
  type: TMessageType;
  createdAt: Date;
  updatedAt: Date;
}

interface IInitialData {
  users: IOtherUsers;
  channels: IChannels;
}

interface IActiveChannelInfo {
  _id: string;
  messages: IMessage[];
  totalMessages: number;
  queue: IMessage[];
}

interface INewGroupCredentials {
  name: string;
  members: string[];
  admins: string[];
}

interface IUpdateGroupCredentials {
  channel_id: string;
  name: string;
  members: string[];
  admins: string[];
}

interface ILeaveGroupCredentials {
  channel_id: string;
}

interface IFetchMessagesCredentials {
  channel_id: string;
  offset: number;
  limit?: number;
  sort?: string;
}

interface ISendMessageCredentials {
  channel_id: string;
  body: string;
}

type TChatState = {
  users: IOtherUsers;
  channels: IChannels;
  activeChannelInfo?: IActiveChannelInfo;
};

enum EChatActions {
  SET_INITIAL_DATA = '@chat/SET_INITIAL_DATA',
  SET_USER_ONLINE = '@chat/SET_USER_ONLINE',
  ADD_OR_UPDATE_CHANNEL = '@chat/ADD_OR_UPDATE_CHANNEL',
  SET_ACTIVE_CHANNEL = '@chat/SET_ACTIVE_CHANNEL',
  REMOVE_CHANNEL = '@chat/REMOVE_CHANNEL',
  ADD_NEW_USER = '@chat/ADD_NEW_USER',
  ADD_MESSAGES = '@chat/ADD_MESSAGES',
  SET_LATEST_MESSAGE = '@chat/SET_LATEST_MESSAGE',
  ADD_TO_MESSAGES_QUEUE = '@chat/ADD_TO_MESSAGES_QUEUE',
  REMOVE_FROM_MESSAGES_QUEUE = '@chat/REMOVE_FROM_MESSAGES_QUEUE',
  SET_LAST_SEEN = '@chat/SET_LAST_SEEN',
  SET_USER_STATUS = '@chat/SET_USER_STATUS',
  SET_USER_THOUGHTS = '@chat/SET_USER_THOUGHTS',
}

interface IChatActionType<T, P> {
  type: T;
  payload: P;
}

type TChatAction =
  | IChatActionType<typeof EChatActions.SET_INITIAL_DATA, IInitialData>
  | IChatActionType<
      typeof EChatActions.SET_USER_ONLINE,
      { _id: string; value: boolean }
    >
  | IChatActionType<typeof EChatActions.ADD_OR_UPDATE_CHANNEL, IChannel>
  | IChatActionType<
      typeof EChatActions.SET_ACTIVE_CHANNEL,
      { _id: string | undefined }
    >
  | IChatActionType<typeof EChatActions.REMOVE_CHANNEL, { channelId: string }>
  | IChatActionType<typeof EChatActions.ADD_NEW_USER, IOtherUser>
  | IChatActionType<
      typeof EChatActions.ADD_MESSAGES,
      { messages: IMessage[]; totalMessages: number; atTop: boolean }
    >
  | IChatActionType<typeof EChatActions.SET_LATEST_MESSAGE, IMessage>
  | IChatActionType<typeof EChatActions.ADD_TO_MESSAGES_QUEUE, IMessage>
  | IChatActionType<typeof EChatActions.REMOVE_FROM_MESSAGES_QUEUE, IMessage>
  | IChatActionType<
      typeof EChatActions.SET_LAST_SEEN,
      {
        channel_id: string;
        user_id: string;
        last_seen: string;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_STATUS,
      {
        user_id: string;
        newStatus: TUserStatus;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_THOUGHTS,
      {
        user_id: string;
        newThoughts: string;
      }
    >;

export type {
  TChatState,
  TChatAction,
  IInitialData,
  IOtherUsers,
  IOtherUser,
  IChannel,
  IChannels,
  IMember,
  IMessage,
  IAttachment,
  INewGroupCredentials,
  IUpdateGroupCredentials,
  ILeaveGroupCredentials,
  IFetchMessagesCredentials,
  ISendMessageCredentials,
};
export { EChatActions };
