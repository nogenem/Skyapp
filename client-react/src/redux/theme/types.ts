type TThemeMode = 'dark' | 'light';

type TThemeState = {
  mode?: TThemeMode;
};

enum EThemeActions {
  MODE_SWITCHED = '@theme/MODE_SWITCHED',
}

interface IThemeActionType<T, P> {
  type: T;
  payload: P;
}

type TThemeAction = IThemeActionType<
  typeof EThemeActions.MODE_SWITCHED,
  TThemeMode
>;

export type { TThemeMode, TThemeState, TThemeAction };
export { EThemeActions };
