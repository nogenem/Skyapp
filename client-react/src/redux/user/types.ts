interface User {
  _id: string;
  token?: string;
}

type UserState = User;

enum EUserActions {
  SIGNED_IN = '@user/SIGNED_IN',
  SIGNED_OUT = '@user/SIGNED_OUT',
}

interface UserActionType<T, P> {
  type: T;
  payload: P;
}

type UserAction =
  | UserActionType<typeof EUserActions.SIGNED_IN, User>
  | UserActionType<typeof EUserActions.SIGNED_OUT, null>;

export type { User, UserState, UserAction };
export { EUserActions };
