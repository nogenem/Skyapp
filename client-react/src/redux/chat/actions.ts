import { Dispatch } from 'redux';

import * as SOCKET_EVENTS from '~/constants/socket_events';
import type { IUser } from '~/redux/user/types';
import api from '~/services/api';
import io from '~/services/io';

import type {
  IChannel,
  IInitialData,
  INewGroupCredentials,
  IOtherUser,
  IUpdateGroupCredentials,
} from './types';
import { EChatActions } from './types';

const setInitialData = (data: IInitialData) => ({
  type: EChatActions.SET_INITIAL_DATA,
  payload: data,
});

const setUserOnline = (_id: string, value: boolean) => ({
  type: EChatActions.SET_USER_ONLINE,
  payload: {
    _id,
    value,
  },
});

const addNewChannel = (channel: IChannel) => ({
  type: EChatActions.ADD_NEW_CHANNEL,
  payload: channel,
});

export const setActiveChannel = (channel_id: string) => ({
  type: EChatActions.SET_ACTIVE_CHANNEL,
  payload: { _id: channel_id },
});

export const connectIo = (user: IUser) => (dispatch: Dispatch) => {
  const instance = io.instance();
  instance.connect({ _id: user._id });

  instance.socket!.on(SOCKET_EVENTS.SOCKET_CONNECT, () => {
    instance.socket!.emit(
      SOCKET_EVENTS.IO_GET_INITIAL_DATA,
      (data: IInitialData) => {
        dispatch(setInitialData(data));
      },
    );

    instance.socket!.on(SOCKET_EVENTS.IO_SIGNIN, (_id: string) => {
      dispatch(setUserOnline(_id, true));
    });
    instance.socket!.on(SOCKET_EVENTS.IO_SIGNOUT, (_id: string) => {
      dispatch(setUserOnline(_id, false));
    });
    instance.socket!.on(
      SOCKET_EVENTS.IO_PRIVATE_CHANNEL_CREATED,
      (channel: IChannel) => {
        dispatch(addNewChannel(channel));
      },
    );
    instance.socket!.on(
      SOCKET_EVENTS.IO_GROUP_CHANNEL_CREATED,
      (channel: IChannel) => {
        dispatch(addNewChannel(channel));
      },
    );
  });
};

export const disconnectIo = () => (dispatch: Dispatch) => {
  io.instance().disconnect();
};

export const startChattingWith = (otherUser: IOtherUser) => (
  dispatch: Dispatch,
) =>
  api.chat.createChannelWith(otherUser).then(({ channel_id }) => {
    dispatch(setActiveChannel(channel_id));
  });

export const createGroupChannel = (credentials: INewGroupCredentials) => (
  dispatch: Dispatch,
) =>
  api.chat.createGroupChannel(credentials).then(({ channel_id }) => {
    dispatch(setActiveChannel(channel_id));
  });

export const updateGroupChannel = (credentials: IUpdateGroupCredentials) => (
  dispatch: Dispatch,
) =>
  api.chat.updateGroupChannel(credentials).then(({ channel_id }) => {
    dispatch(setActiveChannel(channel_id));
  });
