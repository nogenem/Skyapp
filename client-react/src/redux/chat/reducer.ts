import produce from 'immer';
import { createSelector } from 'reselect';

import type { IAppState } from '../store';
import { EChatActions } from './types';
import type { TChatAction, TChatState } from './types';

export const initialState: TChatState = {
  users: {},
  channels: {},
};

export default function chat(
  state = initialState,
  action: TChatAction,
): TChatState {
  return produce(state, draft => {
    switch (action.type) {
      case EChatActions.SET_INITIAL_DATA:
        draft.users = action.payload.users;
        draft.channels = action.payload.channels;
        return draft;
      case EChatActions.SET_USER_ONLINE:
        if (draft.users[action.payload._id]) {
          draft.users[action.payload._id].online = action.payload.value;
        }
        return draft;
      default:
        return draft;
    }
  });
}

// SELECTORS
export const getChat = (state: IAppState) => state.chat || initialState;
export const getUsers = createSelector(getChat, data => data.users);
export const getChannels = createSelector(getChat, data => data.channels);
