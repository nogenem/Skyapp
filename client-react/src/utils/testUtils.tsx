import React from 'react';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import type { IAppState } from '~/redux/store';

export const getMockStore = () => {
  const middlewares = [thunk];
  return configureStore(middlewares);
};

export const getRenderWithRedux = () => {
  const emptyState: Partial<IAppState> = {};
  const mockStore = getMockStore();

  return (
    ui: React.ReactNode,
    initialState: Partial<IAppState> = emptyState,
  ) => {
    const store = mockStore(initialState);
    return {
      ...render(<Provider store={store}>{ui}</Provider>),
    };
  };
};
