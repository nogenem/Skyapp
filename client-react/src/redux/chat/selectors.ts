import { createSelector } from 'reselect';

import type { IAppState } from '../store';
import { selectUserId } from '../user/selectors';
import { initialState } from './reducer';
import { IOtherUser } from './types';
import type { IChannel } from './types';

export const selectChat = (state: IAppState) => state.chat || initialState;

export const selectChatUsers = createSelector(selectChat, data => data.users);

export const selectChatChannels = createSelector(
  selectChat,
  data => data.channels,
);

export const selectChannelFromProps = (
  state: IAppState,
  { channel }: { channel: IChannel },
) => channel || {};

export const selectChatUsersList = createSelector(selectChat, data => {
  const users = Object.values(data.users);
  users.sort((a, b) => a.nickname.localeCompare(b.nickname));
  return users;
});

export const selectChatChannelsList = createSelector(selectChat, data => {
  const channels = Object.values(data.channels);
  channels.sort((a, b) => a.name.localeCompare(b.name));
  channels.sort((a, b) => {
    if (!!a.lastMessage && !!b.lastMessage) {
      return (
        b.lastMessage.updatedAt.getTime() - a.lastMessage.updatedAt.getTime()
      );
    } else if (!!a.lastMessage && !b.lastMessage) {
      return -1;
    } else if (!!b.lastMessage && !a.lastMessage) {
      return 1;
    } else {
      return 0;
    }
  });
  return channels;
});

export const selectChatUsersWithoutChannelList = createSelector(
  selectChat,
  data => {
    const users = Object.values(data.users).filter(user => !user.channel_id);
    users.sort((a, b) => a.nickname.localeCompare(b.nickname));
    return users;
  },
);

export const selectActiveChannelInfo = createSelector(
  selectChat,
  data => data.activeChannelInfo,
);

export const selectActiveChannelId = createSelector(
  selectChat,
  data => data.activeChannelInfo?._id,
);

export const selectActiveChannel = createSelector(
  selectChat,
  data =>
    data.channels[data.activeChannelInfo?._id || ''] as IChannel | undefined,
);

export const selectActiveChannelMessages = createSelector(
  selectChat,
  data => data.activeChannelInfo?.messages,
);

export const selectActiveChannelMessagesQueue = createSelector(
  selectChat,
  data => data.activeChannelInfo?.queue,
);

export const selectActiveChannelTotalMessages = createSelector(
  selectChat,
  data => data.activeChannelInfo?.totalMessages,
);

export const selectOtherUserFromChannel = createSelector(
  [selectChat, selectChannelFromProps],
  (data, channel) => {
    if (!channel.is_group) {
      return data.users[channel.members[channel.other_member_idx || 0].user_id];
    }
    return undefined;
  },
);

export const selectOtherUsersFromActiveChannel = createSelector(
  [selectChat, selectActiveChannel, selectUserId],
  (data, channel, loggedUserId) => {
    const users: IOtherUser[] = [];
    if (channel) {
      channel.members.forEach(member => {
        if (member.user_id !== loggedUserId) {
          users.push(data.users[member.user_id]);
        }
      });
    }
    return users;
  },
);
