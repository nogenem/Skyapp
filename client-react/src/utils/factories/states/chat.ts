import faker from 'faker/locale/en_US';

import { initialState } from '~/redux/chat/reducer';
import type {
  IChannel,
  IChannels,
  IOtherUser,
  IOtherUsers,
  TChatState,
} from '~/redux/chat/types';

import activeChannelInfoFactory from '../models/activeChannelInfo';
import channelFactory from '../models/channel';
import otherUserFactory from '../models/otherUser';

interface IOptions {
  useConstValues: boolean;
  useInitialState: boolean;
  usersLen: number;
  channelsLen: number;
}

export default (
  override?: Partial<TChatState>,
  options?: Partial<IOptions>,
): TChatState => {
  let usersLen = faker.datatype.number({ min: 0, max: 2 });
  if (options?.usersLen !== undefined && options.usersLen >= 0) {
    usersLen = options.usersLen;
  } else if (options?.useConstValues) {
    usersLen = 2;
  }

  let tmpOtherUsers: IOtherUser[] = [];
  if (!options?.useConstValues) {
    tmpOtherUsers = Array.from({ length: usersLen }, () => otherUserFactory());
  } else {
    tmpOtherUsers = Array.from({ length: usersLen }, (_, idx) =>
      otherUserFactory(
        { _id: `other-user-${idx}`, channel_id: `channel-${idx}` },
        { useConstValues: true },
      ),
    );
  }

  const users: IOtherUsers = {};
  tmpOtherUsers.forEach(user => {
    users[user._id] = user;
  });

  let channelsLen = faker.datatype.number({ min: 0, max: 2 });
  if (options?.channelsLen !== undefined && options.channelsLen >= 0) {
    channelsLen = options.channelsLen;
  } else if (options?.useConstValues) {
    channelsLen = 2;
  }

  let tmpChannels: IChannel[] = [];
  if (!options?.useConstValues) {
    tmpChannels = Array.from({ length: channelsLen }, () => channelFactory());
  } else {
    tmpChannels = Array.from({ length: channelsLen }, (_, idx) =>
      channelFactory({ _id: `channel-${idx}` }, { useConstValues: true }),
    );
  }

  const channels: IChannels = {};
  tmpChannels.forEach(channel => {
    channels[channel._id] = channel;
  });

  let activeChannelInfo =
    Math.random() < 0.5 ? undefined : activeChannelInfoFactory();
  if (options?.useConstValues) activeChannelInfo = undefined;

  return options?.useInitialState
    ? initialState
    : ({
        users,
        channels,
        activeChannelInfo,
        ...(override || {}),
      } as TChatState);
};

export type { IOptions };
