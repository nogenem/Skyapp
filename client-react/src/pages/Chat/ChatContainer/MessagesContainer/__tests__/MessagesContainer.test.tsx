import React from 'react';

import { IMessage } from '~/redux/chat/types';
import FACTORIES from '~/utils/factories';
import { getRenderWithRedux } from '~/utils/testUtils';

import { MessagesContainer } from '../index';

const renderWithRedux = getRenderWithRedux();

describe('MessagesContainer', () => {
  it('renders correctly', () => {
    const messages: IMessage[] = [
      FACTORIES.models.message(
        {
          _id: 'message-1',
          body: 'message 1',
        },
        { useConstValues: true },
      ),
    ];
    const messagesQueue: IMessage[] = [
      FACTORIES.models.message(
        {
          _id: 'queue-message-1',
          body: 'queued message',
        },
        { useConstValues: true },
      ),
    ];
    const loggedUser = FACTORIES.models.user({}, { useConstValues: true });

    const { container, getByText } = renderWithRedux(
      <MessagesContainer
        messages={messages}
        messagesQueue={messagesQueue}
        loggedUser={loggedUser}
        users={{}}
        onScrollTop={async () => {}}
        changeEditingMessage={(messageId: string) => {}}
        onDeleteMessage={(messageId: string) => {}}
      />,
    );

    expect(container).toMatchSnapshot();
    expect(getByText(messages[0].body as string)).toBeInTheDocument();
    expect(getByText(messagesQueue[0].body as string)).toBeInTheDocument();
  });
});
