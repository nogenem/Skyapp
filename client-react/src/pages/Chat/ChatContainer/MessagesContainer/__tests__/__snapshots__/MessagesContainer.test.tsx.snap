// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesContainer renders correctly 1`] = `
<div>
  <div
    class="makeStyles-container-1"
  >
    <div
      class="makeStyles-date-21 makeStyles-fancy-22"
    >
      04/12/2021
    </div>
    <div
      class="makeStyles-system_message_wrapper-13"
    >
      <div
        class="MuiPaper-root makeStyles-system_messages-14 MuiPaper-elevation1 MuiPaper-rounded"
      >
        message 1
      </div>
    </div>
    <div
      class="makeStyles-messageWrapper-8 makeStyles-messageWrapperFromThem-10"
    >
      <div
        class="MuiPaper-root makeStyles-messageFromThem-3 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <p
          class="MuiTypography-root makeStyles-text_message-19 MuiTypography-body2"
        >
          queued message
        </p>
      </div>
      <div
        class="MuiCircularProgress-root makeStyles-loading_icon-11 MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
        title="Messages:Sending the message"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20"
            stroke-width="4"
          />
        </svg>
      </div>
    </div>
    <button
      aria-controls="msg-menu-channel-1"
      aria-haspopup="true"
      aria-label="Common:Options"
      class="MuiButtonBase-root MuiIconButton-root"
      style="position: absolute; padding: 0px; top: -1px; left: -1px; display: none;"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
